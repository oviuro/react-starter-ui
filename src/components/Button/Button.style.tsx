import * as React from 'react'
import {css} from 'styled-components'
import {ButtonProps} from "./Button";
import {mixins, vars} from "./Button.theme";

const { colors } = vars

export const button = ({}: ButtonProps) => css`
	appearance: none;
	border: 0;
	cursor: pointer;
	display: inline-block;
	-webkit-font-smoothing: antialiased;
	line-height: 1;
	text-align: center;
	text-decoration: none;
	user-select: none;
	vertical-align: middle;
	white-space: nowrap;
	outline: none;
	background: none;
	border-radius: ${vars.borderRadius};
	padding: ${vars.padding};
	transition: ${vars.transition};
	
	&:disabled {
		cursor: not-allowed;
		${disabled}
	}
	
	${props}
`

const disabled = css`
	color: ${colors.disabled};
	border: ${vars.border};
	background-color: ${colors.disabled_background};
	&:hover {
		color: ${colors.disabled_hover};
		border: ${vars.border};
		background-color: ${colors.disabled_background_hover};
	}
`

const props = (props: ButtonProps) => css`
	${!props.text && mixins.colored(colors.secondary, 'black')}
	${props.primary && mixins.colored(colors.primary)}
	${props.info && mixins.colored(colors.info)}
	${props.success && mixins.colored(colors.success)}
	${props.warning && mixins.colored(colors.warning)}
	${props.error && mixins.colored(colors.error)}
	${props.ghost && ghost}
	${props.dashed && dashed}
	${props.text && text}
	${props.full && 'width: 100%'}
`

const ghost = css`
	background-color: transparent;
	border: ${vars.border};
	color: ${colors.text};
	&:hover {
		background-color: transparent;
		border-color: ${colors.primary};
		color: ${colors.primary};
	}
	${mixins.focus(colors.primary)};
`

const dashed = css`
	${ghost};
	border: 1px dashed ${colors.border};
`

const text = css`
	color: ${colors.text};
	&:hover {
		color: ${colors.primary}
	}
`
