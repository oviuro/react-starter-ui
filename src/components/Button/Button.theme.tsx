import {shade, tint, transparentize} from 'polished'
import {colors} from '../../helpers/colors'
import {settings} from '../../helpers/settings'
import {css} from "styled-components";

export class vars {
	static borderRadius = settings.borderRadius
	static padding = `${settings.smallSpacing} ${settings.baseSpacing}`
	static transition = `all ${settings.transition}`
	static border = settings.border
	static colors = {
		primary: colors.primary,
		secondary: colors.secondary,
		info: colors.info,
		success: colors.success,
		warning: colors.warning,
		error: colors.error,
		disabled: shade(0.7, colors.border),
		disabled_hover: shade(0.8, colors.border),
		disabled_background: tint(0.5, colors.border),
		disabled_background_hover: tint(0.7, colors.border),
		text: colors.text,
		border: colors.border
	}

}

export class mixins {
	static colored = (color: string, textColor: string = 'white') => css`
		color: ${textColor};
		background-color: ${color};
		&:hover {
			background-color: ${shade(0.9, color)}
		}
		${mixins.focus(color)}
	`
	static focus = (color: string) => css`
		&:focus { box-shadow: 0 0 0 3px ${ transparentize(0.8, color) } };
	`
}