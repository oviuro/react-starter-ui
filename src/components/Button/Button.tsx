import * as React from 'react'
import {HTMLProps, StatelessComponent} from "react";

type ButtonProps = HTMLProps<HTMLButtonElement> & {
	primary?: boolean
}

const ButtonSFC: StatelessComponent<ButtonProps> = ({children, ...props}) => (
	<button {...props}>{children}</button>
)

export const Button = ButtonSFC
