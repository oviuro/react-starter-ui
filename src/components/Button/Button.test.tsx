import * as React from 'react'
import {shallow} from  'enzyme'
import {Button} from './Button'
import styled from 'styled-components'
import * as renderer from 'react-test-renderer'
import 'jest-styled-components'

describe('Button component', () => {
	describe('ui', () => {
		it('needs to render without props', () => {
			const button = shallow(<Button />)
			expect(button).toBeDefined()
		})

		it('needs to render with children', () => {
			const button = shallow(<Button>Button Text</Button>)
			expect(button.children().text()).toEqual('Button Text')
		})

		it('not render with children when not passed', () => {
			const button = shallow(<Button />)
			expect(button.children()).toHaveLength(0)
		})

		it('needs to render with primary', () => {
			const button = shallow(<Button primary/>)
			expect(button.props().primary).toBeTruthy()
		})

		it('not render with primary when not passed', () => {
			const button = shallow(<Button />)
			expect(button.props().primary).toBeFalsy()
		})

		it('needs to render with correct theme', () => {})

		it('needs to match snapshot for styles', () => {
			const tree = renderer.create(<Button primary/>).toJSON()
			expect(tree).toMatchSnapshot()
		})
	})
})
