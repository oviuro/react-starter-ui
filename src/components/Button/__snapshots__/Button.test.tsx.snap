// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button component ui needs to match snapshot for styles 1`] = `
<button
  primary={true}
/>
`;
